AVRO-1695: Ruby support for logical types revisited

tjwp opened the pull request:
This is a continuation of the changes originally submitted in https://github.com/apache/avro/pull/44 for https://issues.apache.org/jira/browse/AVRO-1695.

Original description:

```
Add Ruby support for logical types

Only logical types that map directly to Ruby standard library types are supported:

* date maps to Date;
* timestamp-millis and timestamp-micros map to Time.

The remaining types are difficult to cleanly map to Ruby types, so I've refrained from doing so.   
Furthermore, there's no direct support for plugging in custom mappers – I'm unsure if this is     
needed.
```

I've tried to add address the feedback from the original PR, primarily pushing some of the implementation into `Avro::Schema.real_parse` and ensuring that the Ruby tests cover all the same cases as the Java conversion tests.

One change I've made here is to have `Avro::Schema#logical_type` return the name of the logical type.

For now I've preserved the original changes as a separate commit, but I can rebase the changes if this gets accepted.

CC @dasch

dasch commented:
:+1:

tjwp commented:
Updated Manifest and changes to ensure that `TimestampMicros` works with `DateTime`.

tjwp commented:
@rdblue You reviewed the original pull request for this support in ruby. Is there anything that we can do to help push this along?

rdblue commented:
Sorry for the delay. I'll review this soon. I'm planning on making some time next weekend (17-18 Dec).

busbey commented:
If you wouldn't mind rebasing to current master, I'll take a look at this my next time around (~2/7)

tjwp commented:
@busbey I've rebased this, and also squashed the commits. I'll also rebase https://github.com/apache/avro/pull/170, another ruby PR, against current master.

tjwp commented:
@spacharya I noticed that you recently looked at another Ruby PR to potentially be merged soon. Would you mind taking a look at this one too? We've been using this code in production for a while.

busbey commented:
I thought of another question: anyone know if we have any cross-language compatibility tests in `share/ ` for logical types yet? If we do, I'd want to make sure we're good here. If not, I just want to make sure we have a follow-on somewhere for it.

tjwp commented:
I'll resolve the conflicts on this tomorrow. Also, happy to help with the ruby suite tests if there's anything I can do there.

tjwp commented:
Rebased on the current master.

tjwp commented:
@busbey I took a look through `share/` and there does not seem to be anything for logical types yet.

tjwp commented:
@busbey Any chance of getting this merged prior to the next 1.8.2 RC?

Fokko commented:
@tjwp Can you rebase onto master?

tjwp commented:
@Fokko rebased and tested with:

SUCCESS rvm: 1.9.3-p551
SUCCESS rvm: 2.0.0-p648
SUCCESS rvm: 2.1.10
SUCCESS rvm: 2.2.8
SUCCESS rvm: 2.3.8
SUCCESS rvm: 2.4.5
SUCCESS rvm: 2.5.3

tjwp commented:
Rebased again with a fix for the ruby tests and the additional change referenced here: https://github.com/salsify/avro-patches/pull/18#issue-202225742

Fokko commented:
@tjwp Travis is still not happy, can you take a peek?
```
.................../testptch/unknown/lang/ruby/lib/avro/schema.rb:318: warning: too many arguments for format string
..........................F
===============================================================================
Failure: test_timestamp_micros_long_conversion(TestLogicalTypes)
/testptch/unknown/lang/ruby/test/test_logical_types.rb:83:in `test_timestamp_micros_long_conversion'
     80: 
     81:     now = Time.now.utc
     82: 
  => 83:     assert_equal now, type.decode(type.encode(now))
     84:     assert_equal 1432849613221843, type.encode(Time.utc(2015, 5, 28, 21, 46, 53, 221843))
     85:     assert_equal 1432849613221843, type.encode(DateTime.new(2015, 5, 28, 21, 46, 53.221843))
     86:     assert_equal Time.utc(2015, 5, 28, 21, 46, 53, 221843), type.decode(1432849613221843)
<2018-11-07 20:38:27 UTC> expected but was
<2018-11-07 20:38:27 UTC>

diff:
  2018-11-07 20:38:27 UTC
===============================================================================
```

Fokko commented:
Thanks @tjwp
